### PR TITLE
Improvements to docs on custom implementations

### LINKED ISSUE #2005: State Changes from List to Tensor After Calling Compute()
## 🐛 Bug

I was logging a custom Metric result using the logger provided by LightningModule. The codes runs on 2 GPUs with ddp. While a list state whose `dist_reduce_fx` is "cat" was synchronized by calling `compute()` , I realized the behavior in `training_step()` and `validation_step()` were different.

The state in `training_step()` where on_step=True and on_epoch=False (default setting for training) after calling `compute()` was always a list. However, in `validation_step()` where on_step=False and on_epoch=True (default setting for validation), the state became a Tensor. Such behavior is not explained in the doc.

### To Reproduce

<details>
  <summary>Metric and Logging Results</summary>

**Codes**
```python
import logging

import torch
from torchmetrics import Metric
from torchmetrics.functional.retrieval import retrieval_normalized_dcg


class MyMetric(Metric):
    def __init__(self, top_k):
        super().__init__()
        self.top_k = top_k
        self.add_state("state_1", default=[], dist_reduce_fx="cat")

    def update(self, preds: Tensor, target: Tensor):
        assert preds.shape == target.shape
        self.state_1 += [self._metric(p, t) for p, t in zip(preds, target)]

    def compute(self):
        if isinstance(self.state_1, list):
            logging.warning(f"self.state_1.size(): {len(self.state_1)}")
            return torch.stack(self.state_1).mean()
        logging.warning(f"self.state_1.size(): {self.state_1.size()}")
        return self.state_1.mean()

    def _metric(self, preds: Tensor, target: Tensor):
        return retrieval_normalized_dcg(preds, target, k=self.top_k).float()
```

**Log Outputs**
Training:
`WARNING:self.state_1.size(): 40`

Validation:
`WARNING:self.state_1.size(): torch.Size([80])`

</details>

### Expected behavior

The state is always what it is defined in add_state(), i.e., a list.

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source):
0.10.3 (conda-forge)
- Python & PyTorch Version (e.g., 1.0):
3.10.12
- Any other relevant information such as OS (e.g., Linux):
pytorch: 1.13.1
pytorch-lightning: 1.9.4
Linux: 5.15.0-78-generic

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2005 
Trying to give the documentation on creating custom implementation an update.
As mentioned in the issue, some details especially regarding list metric states are missing.
Also did a bit of an update to the utilities section.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2061.org.readthedocs.build/en/2061/

### CODE REVIEW

File: docs/source/pages/implement.rst [hunk 1]
```diff
@@ -17,36 +27,106 @@ methods:
...
+                raise ValueError("preds and target must have the same shape")
 
             self.correct += torch.sum(preds == target)
             self.total += target.numel()
 
-        def compute(self):
+        def compute(self) -> Tensor:
             return self.correct.float() / self.total
 
+A few important things to note:
```
Comment (Borda) [inline docs/source/pages/implement.rst:62]: maybe use `.. note::`?

File: docs/source/pages/implement.rst [hunk 2]
```diff
@@ -17,36 +27,106 @@ methods:
...
+        def compute(self):
+            preds = dim_zero_cat(preds)
+            target = dim_zero_cat(target)
+            r_preds, r_target = _rank_data(preds), _rank_dat(target)
+            preds_diff, target_diff  = r_preds - r_preds.mean(0), r_target - r_target.mean(0)
+            cov = (preds_diff * target_diff).mean(0)
+            preds_std = torch.sqrt((preds_diff * preds_diff).mean(0))
+            target_std = torch.sqrt((target_diff * target_diff).mean(0))
+            corrcoef = cov / (preds_std * target_std + eps)
+            return torch.clamp(corrcoef, -1.0, 1.0)
```
Comment (Borda) [inline docs/source/pages/implement.rst:108]:
```suggestion
        def compute(self):
            # parse inputs
            preds = dim_zero_cat(preds)
            target = dim_zero_cat(target)
            # some intermediate computation...
            r_preds, r_target = _rank_data(preds), _rank_dat(target)
            preds_diff = r_preds - r_preds.mean(0)
            target_diff = r_target - r_target.mean(0)
            cov = (preds_diff * target_diff).mean(0)
            preds_std = torch.sqrt((preds_diff * preds_diff).mean(0))
            target_std = torch.sqrt((target_diff * target_diff).mean(0))
            # finalize the computations
            corrcoef = cov / (preds_std * target_std + eps)
            return torch.clamp(corrcoef, -1.0, 1.0)
```
have it easier to read

File: docs/source/pages/implement.rst [hunk 3]
```diff
@@ -17,36 +27,106 @@ methods:
...
+
+A few important things to note for this example:
+
+* When working with list states, the ``dist_reduce_fx`` argument to ``add_state`` should be set to ``"cat"`` to
+  concatenate the list of tensors across batches.
 
-Additionally you may want to set the class properties: `is_differentiable`, `higher_is_better` and
-`full_state_update`. Note that none of them are strictly required for the metric to work.
+* The ``update`` method when working with list states should append the batch states to the list.
```
Comment (Borda) [inline docs/source/pages/implement.rst:115]:
```suggestion
* When working with list states, The ``update(...)`` method should append the batch states to the list.
```

File: docs/source/pages/implement.rst [hunk 4]
```diff
@@ -65,8 +145,12 @@ Additionally you may want to set the class properties: `is_differentiable`, `hig
         # batch states are independent and we will optimize the runtime of 'forward'
         full_state_update: bool = True
 
-Finally, from torchmetrics v1.0.0 onwards, we also support plotting of metrics through the `.plot` method. By default
-this method will raise `NotImplementedError` but can be implemented by the user to provide a custom plot for the metric.
+**************
+Plot interface
+**************
+
+From torchmetrics v1.0.0 onwards, we also support plotting of metrics through the `.plot` method. By default this method
```
Comment (Borda) [inline docs/source/pages/implement.rst:152]:
```suggestion
From torchmetrics v1.0.0 onwards, we also support plotting of metrics through the ``.plot()`` method. By default this method
```

**Review by Borda — APPROVED** [review]

File: docs/source/pages/implement.rst [hunk 5]
```diff
@@ -194,8 +288,8 @@ and tests gets formatted in the following way:
   5. (optional) If your metric raises any exception, please add tests that showcase this.
 
   .. note::
-    The `test file for accuracy <https://github.com/Lightning-AI/torchmetrics/blob/master/tests/unittests/classification/test_accuracy.py>`_ metric
-    shows how to implement such tests.
+    The `test file for mse <https://github.com/Lightning-AI/torchmetrics/blob/master/tests/unittests/regression/test_mean_error.py>`_
```
Comment (Borda) [inline docs/source/pages/implement.rst:291]:
```suggestion
    The `test file for MSE <https://github.com/Lightning-AI/torchmetrics/blob/master/tests/unittests/regression/test_mean_error.py>`_
```

**Review by justusschock — APPROVED** [review]
